### PR TITLE
[DTensor] Fix _StridedShard to Replicate padding issue

### PR DESCRIPTION
**Pull request by zpcore**
Fix `_StridedShard._to_replicate_tensor` incorrectly unpad the tensors after all gather. 

`_StridedShard._to_replicate_tensor` previously assumes the padding can only happen at the end of the gathered tensor. However, due to the split_factor, the padding can occur on different chunks. E.g., 

```
import torch
from torch.distributed.tensor.placement_types import _StridedShard
a = torch.arange(9) # 1D tensor with 9 elements
SS = _StridedShard(0, split_factor=2)
num_chunks = 4
output = SS._split_tensor(a, num_chunks)
```
We can see the output splitted tensor contains 4 chunks:
`[0, 1, 5], [2, 3, 6], [4, 7], [8]`
When we try pad before allgather, the 4 chunks become:
`[0, 1, 5], [2, 3, 6], [4, 7, Pad], [8, Pad, Pad]`
After allgather, we have:
`[0, 1, 5, 2, 3, 6, 4, 7, Pad, 8, Pad, Pad]`
Original assumption is that there are total 3 paddings (based on max_chunk_size * num_splits), so we trim the last 3 elements, and the final result becomes:
`[0, 1, 5, 2, 3, 6, 4, 7, Pad]`
Now we keep track of the padding and will only select the nonpadding elements after allgather.

See below picture for illustration on how the StridedShard data get reshuffled back after allgather (mesh dim 0 size is 4) :
image
code accompanying diagram for index select usage:
```
        padded_positions = []
        for i, shard in enumerate(sharded_indices):
            base_offset = i * max_chunk_size
            positions = base_offset + torch.arange(
                len(shard), device=local_tensor.device
            )
            padded_positions.append(positions)

        permutation = torch.cat(sharded_indices)
        # Choose the position by skipping padding indices from
        # replicate_tensor_permuted_padded.
        select_positions = torch.cat(padded_positions)

        inv_permutation = torch.argsort(permutation)
        select_indices = select_positions.index_select(0, inv_permutation)

        replicate_tensor = torch.index_select(
            replicate_tensor_permuted_padded, self.dim, select_indices
        )

```
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #171338
* #171219
* #171178
* #171023
* #170915
* #172266
* #171337
* __->__ #170914

### CODE REVIEW

File: test/distributed/tensor/test_utils.py [hunk 1]
```diff
@@ -1496,6 +1498,128 @@ def tearDown(self):
...
+                        mesh,
+                        0,
+                        a.shape,
+                    )
+                    p = _StridedShard(0, split_factor=split_factor)
+                    a_dt_after_to_replicate = p._to_replicate_tensor(
+                        a_dt.to_local(), mesh, 0, logical_shape
+                    )
+                    b_dt = distribute_tensor(a, mesh, [Replicate()], src_data_rank=None)
+                except Exception:
```
Comment (wconstab) [inline test/distributed/tensor/test_utils.py]: uhh.. i don't think you should have a catchall here - how do i know that any of the test cases are actually running instead of throwing an unrelated error and skipping?

File: torch/distributed/tensor/placement_types.py [hunk 2]
```diff
@@ -741,19 +742,30 @@ def _to_replicate_tensor(
         if isinstance(replicate_tensor_permuted_padded, funcol.AsyncCollectiveTensor):
             replicate_tensor_permuted_padded = replicate_tensor_permuted_padded.wait()
 
-        if replicate_tensor_permuted_padded.shape[self.dim] > logical_dim_size:
-            replicate_tensor_permuted = unpad_tensor(
-                replicate_tensor_permuted_padded,
-                self.dim,
-                replicate_tensor_permuted_padded.shape[self.dim] - logical_dim_size,
+        # After all_gather, the tensor is [chunk0, chunk1 (may padded), ...].
```
Comment (wconstab) [inline torch/distributed/tensor/placement_types.py:891]: I read your PR desc and this comment, and my reaction was "don't we follow torch.chunk semantics, meaning we keep padding only on the end"?  It might help to remind us where the logic lives for chunking.  


I am again confused by what the difference is between `_split_tensor` and `_shard_tensor` method on StridedShard class.

**Review by wconstab — COMMENTED** [review]
can you update the PR desc to show 2 things

1) tell us what the mesh dim size is
2) run python interpreter to simulate the torch.chunk behavior directly? IIUC the way StridedShard got created via _split_tensor, it should call chunk twice.  But i am not sure I understand how this results in the padding combination you claimed.

File: test/distributed/tensor/test_utils.py (hunk 1 above)
Comment (zpcore) [reply test/distributed/tensor/test_utils.py]: Good point, the exception can be removed actually. Fixed.

File: torch/distributed/tensor/placement_types.py (hunk 2 above)
Comment (zpcore) [reply torch/distributed/tensor/placement_types.py:891]: For each chunk before we perform allgather, the padding can exist at the end of each chunk. Then after all gather those chunks into one tensor, we can have paddings in between.

Note that this is not an issue for normal `Shard`, because chunks from Shard can only have paddings in the last chunk. After all gahter, still the padding will be at the end of the gathered tensor.

> I am again confused by what the difference is between _split_tensor and _shard_tensor method

To be simple, we can think of `_split_tensor` the same as `_shard_tensor` when we set _shard_tensor(src_data_rank=None). This will skip checking all shards after split should be equal. In fact, I think we should allow `_shard_tensor` to accept unequal splits.

File: torch/distributed/tensor/placement_types.py [hunk 3]
```diff
@@ -803,19 +804,30 @@ def _to_replicate_tensor(
...
+            base_offset = i * max_chunk_size
+            positions = base_offset + torch.arange(
+                len(shard), device=local_tensor.device
             )
-        else:
-            replicate_tensor_permuted = replicate_tensor_permuted_padded
+            padded_positions.append(positions)
 
         permutation = torch.cat(sharded_indices)
+        # Choose the position by skipping padding indices from
```
Comment (wconstab) [inline torch/distributed/tensor/placement_types.py:908]: naively it looks like permutation (torch.cat(sharded_indices)) should be equivalent to torch.arange(n), which then makes argsort and therefore inv_permutation very uninteresting.  only after a few reads did I realize that becuase we split the non-flattened indices tensor, we get an interesting pattern here. might be worth adding a comment above permutation "permutation ends up containing strided indices becuase we create it by chunking over a particular dimension of an N-D shaped arange tensor."

File: torch/distributed/tensor/placement_types.py [hunk 4]
```diff
@@ -803,19 +804,30 @@ def _to_replicate_tensor(
...
-                self.dim,
-                replicate_tensor_permuted_padded.shape[self.dim] - logical_dim_size,
+        # After all_gather, the tensor is [chunk0, chunk1 (may padded), ...].
+        # Each chunk may have padding at the end. Use a single index_select to
+        # both extract non-padding data and reorder to the original positions.
+        #
+        # Build select_indices where select_indices[original_pos] = position in
+        # the padded tensor that holds the element for original_pos.
+        padded_positions = []
+        for i, shard in enumerate(sharded_indices):
```
Comment (wconstab) [inline torch/distributed/tensor/placement_types.py:898]: I like the diagram you added to the PR desc but I want code examples too.  I'd like an example of what padded_positions looks like and what permutation looks like.  i think it makes sense to take the 9 element example and make it a running example through this code

**Review by wconstab — COMMENTED** [review]
lgtm. it took me a while to understand the code though..

File: test/distributed/tensor/test_utils.py [hunk 5]
```diff
@@ -1496,6 +1498,124 @@ def tearDown(self):
             pass
 
 
+class TestStridedShardCollectiveOpUtils:
+    from collections import namedtuple
+
+    ShardConfig = namedtuple("ShardConfig", ["mesh_dim", "split_factor"], defaults=(1,))
+
+    def _range_tensor(self, *shape) -> torch.Tensor:
```
Comment (weifengpy) [inline test/distributed/tensor/test_utils.py]: who calls this `_range_tensor`?

File: torch/distributed/tensor/placement_types.py [hunk 6]
```diff
@@ -825,21 +837,19 @@ def _to_replicate_tensor(
...
         rank: RankType,
         return_first_offset: bool = True,
     ) -> tuple[int, int | list[int]]:
-        return _StridedShard.local_shard_size_and_offset(
-            self, curr_local_size, num_chunks, rank, return_first_offset
+        return self.local_shard_size_and_offset(
+            curr_local_size, num_chunks, rank, return_first_offset
         )
 
-    @staticmethod
```
Comment (weifengpy) [inline torch/distributed/tensor/placement_types.py:840]: thanks for cleaning this up

**Review by weifengpy — APPROVED** [review]

File: test/distributed/tensor/test_utils.py (hunk 5 above)
Comment (zpcore) [reply test/distributed/tensor/test_utils.py]: Will be used in future PRs.  I have relocated it to that PR.

File: torch/distributed/tensor/placement_types.py (hunk 4 above)
Comment (zpcore) [reply torch/distributed/tensor/placement_types.py:898]: Added the AI generated example, which is the same as in the PR description picture.

File: torch/distributed/tensor/placement_types.py (hunk 3 above)
Comment (zpcore) [reply torch/distributed/tensor/placement_types.py:908]: Hah, the `sharded_indices` is the key to permute _StridedShard order back. This will also be used in all other `_StridedShard <-> Shard`, `_StridedShard <-> Paritial` conversions.

**Review by wconstab — APPROVED** [review]
lgtm!
